test/raft/distributed_test.cc

```diff
@@ -530,5 +533,85 @@ struct persistence : public raft::persistence {
     }
 };
 
+struct failure_detector : public raft::failure_detector, public server_set {
+    raft::logical_clock _clock;
+
+    // The set of known servers, used to broadcast heartbeats.
+    // The second element of the pair denotes an expiration time after which we remove the server from this set.
+    // Most servers won't have an expiration time; it is assigned only to servers from which we receive a message
+    // without having them added locally with `add_server`.
+    std::unordered_map<raft::server_id, std::optional<raft::logical_clock::time_point>> _known;
+
+    // The last time we received a heartbeat from a server.
+    std::unordered_map<raft::server_id, raft::logical_clock::time_point> _last_heard;
+
+    // The last time we sent a heartbeat.
+    raft::logical_clock::time_point _last_beat;
+
+    using send_heartbeat_t = std::function<void(raft::server_id dst)>;
+
+    send_heartbeat_t _send_heartbeat;
+
+    failure_detector(send_heartbeat_t f)
+        : _send_heartbeat(std::move(f))
+    {}
+
+    void receive_heartbeat(raft::server_id src) {
+        // TODO: make it adjustable
+        static const raft::logical_clock::duration _known_gc_threshold = 30s;
+        auto expiration = _clock.now() + _known_gc_threshold;
+
+        auto [it, inserted] = _known.emplace(src, expiration);
+        if (!inserted && it->second && it->second < expiration) {
+            it->second = expiration;
+        }
+
+        _last_heard[src] = std::max(_clock.now(), _last_heard[src]);
+    }
+
+    void advance(raft::logical_clock::duration d) {
+        _clock.advance(d);
+
+        // TODO: make it adjustable
+        static const raft::logical_clock::duration _heartbeat_period = 100ms;
+        if (_clock.now() - _last_beat > _heartbeat_period) {
+            for (auto& [dst, _] : _known) {
+                _send_heartbeat(dst);
+            }
+            _last_beat = _clock.now();
+        }
+
+        // Garbage-collect _known servers
+        std::vector<raft::server_id> to_erase;
+        for (auto& [dst, gc_time_opt] : _known) {
+            if (gc_time_opt && _clock.now() > *gc_time_opt) {
+                to_erase.push_back(dst);
+            }
+        }
+
+        for (auto dst : to_erase) {
+            _known.erase(dst);
+        }
+    }
+
+    virtual void add_server(raft::server_id id) override {
+        auto [it, inserted] = _known.emplace(id, std::nullopt);
+        if (!inserted && it->second) {
+            it->second = std::nullopt;
+        }
+    }
+
+    virtual void remove_server(raft::server_id id) override {
+        _known.erase(id);
+    }
+
+    bool is_alive(raft::server_id id) override {
+        // TODO: make it adjustable
+        static const raft::logical_clock::duration _convict_threshold = 1s;
```

Comment:
The clock is logical (just a counter) but you compare it to 1 second. How does it work?

Reply:
Well I guess it's (`1s / raft::logical_clock::period`) clock ticks.

Reply:
May be. I just do not know and am scared :)

Reply:
(which does not make any sense)

Reply:
Since GH did not invent the concept of thread yet (when they will they surely will have a huge PR about exiting new feature) I have no idea if your last reply is to me or to your own comment :)

Reply:
Using `1s` does not make sense here